src/core/net/nat64_translator.cpp

```diff
@@ -146,23 +150,29 @@ Translator::Result Translator::TranslateFromIp6(Message &aMessage)
     ip4Header.Clear();
     ip4Header.InitVersionIhl();
     ip4Header.SetSource(mapping->mIp4);
-    ip4Header.GetDestination().ExtractFromIp6Address(mNat64Prefix.mLength, ip6Header.GetDestination());
-    ip4Header.SetTtl(ip6Header.GetHopLimit());
+    ip4Header.GetDestination().ExtractFromIp6Address(mNat64Prefix.mLength, ip6Header.GetDestinationAddress());
+    ip4Header.SetTtl(ip6Header.GetIpHopLimit());
     ip4Header.SetIdentification(0);
 
-    switch (ip6Header.GetNextHeader())
+    switch (ip6Header.GetIpProto())
     {
     case Ip6::kProtoUdp:
         ip4Header.SetProtocol(Ip4::kProtoUdp);
+        udpHeader = ip6Header.GetUdpHeader();
+        udpHeader.SetSourcePort(srcPortOrId);
+        aMessage.Write(0, udpHeader);
```

Comment:
We can consider adding new helper methods to the `Ip6::Headers` class, for example, a non-`const` flavor of `GetUdpHeader()` or `GetTcpHeader()`?

This can help simplify the code. We can then update `ip6Headers` directly and will not need to have a local `udpHeader` and copy `udpHeader = ip6Headers.GetUdpHeader()`.

What do you think?

Reply:
Yes, I like that, it will make the code much cleaner. Should I create a separate PR for this?

Reply:
Submitted https://github.com/openthread/openthread/pull/11151 to address this improvement.